### PR TITLE
Introduce libuzfs and uzfs utility

### PR DESCRIPTION
**Pull request by hpingfs**
libuzfs is a library based on libzpool, providing higher level
abstraction for DMU/SPA/DSL/ZAP.

uzfs is a cmd invoking libuzfs for test purpose.

Signed-off-by: Ping Huang <huangping@smartx.com>

### CODE REVIEW

File: cmd/uzfs/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,14 @@
+include $(top_srcdir)/config/Rules.am
+
+# Unconditionally enable ASSERTs
+AM_CPPFLAGS += -DDEBUG -UNDEBUG -DZFS_DEBUG
+
+sbin_PROGRAMS = uzfs
```
Comment (kylezh) [inline cmd/uzfs/Makefile.am:6]:
```suggestion
bin_PROGRAMS = uzfs
```

File: cmd/uzfs/uzfs.c [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+	 * See comments at end of main().
+	 */
+	if (getenv("ZFS_ABORT") != NULL) {
+		(void) printf("dumping core by request\n");
+		abort();
+	}
+
+	exit(requested ? 0 : 2);
+}
+
```
Comment (kylezh) [inline cmd/uzfs/uzfs.c:186]: additional blank line

File: lib/libuzfs/libuzfs.c [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+			ASSERT0(err);
+		}
+	}
+	return (0);
+}
+
+void
+libuzfs_dataset_destroy(const char *dsname)
+{
+	(void) dmu_objset_find(dsname, libuzfs_objset_destroy_cb, NULL,
```
Comment (kylezh) [inline lib/libuzfs/libuzfs.c:475]: just a find?

File: cmd/uzfs/uzfs.c [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+	return (err);
+}
+
+int
+uzfs_object_delete(int argc, char **argv)
+{
+	int err = 0;
+	char *dsname = argv[1];
+	uint64_t obj = atoi(argv[2]);
+	printf("destroying object %s:%ld\n", dsname, obj);
```
Comment (kylezh) [inline cmd/uzfs/uzfs.c:415]:
```suggestion

	printf("destroying object %s:%ld\n", dsname, obj);
```

File: cmd/uzfs/uzfs.c [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+	printf("\tfill_count: %ld\n", doi->doi_fill_count);
+}
+
+int
+uzfs_object_stat(int argc, char **argv)
+{
+	int err = 0;
+	char *dsname = argv[1];
+	uint64_t obj = atoi(argv[2]);
+	printf("stating object %s:%ld\n", dsname, obj);
```
Comment (kylezh) [inline cmd/uzfs/uzfs.c:492]:
```suggestion

	printf("stating object %s:%ld\n", dsname, obj);
```

File: cmd/uzfs/uzfs.c [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+
+	libuzfs_dataset_handle_t *dhp = libuzfs_dataset_open(dsname);
+	if (!dhp) {
+		printf("Failed to open dataset: %s\n", dsname);
+		return (-1);
+	}
+
+	dmu_object_info_t doi;
+	memset(&doi, 0, sizeof (doi));
+	err = libuzfs_object_stat(dhp, obj, &doi);
```
Comment (kylezh) [inline cmd/uzfs/uzfs.c:502]:
```suggestion

	err = libuzfs_object_stat(dhp, obj, &doi);
```

File: cmd/uzfs/uzfs.c [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+
+	libuzfs_dataset_handle_t *dhp = libuzfs_dataset_open(dsname);
+	if (!dhp) {
+		printf("Failed to open dataset: %s\n", dsname);
+		return (-1);
+	}
+
+	libuzfs_object_list(dhp);
+
+	printf("Listed object: %s\n", dsname);
```
Comment (kylezh) [inline cmd/uzfs/uzfs.c:527]:
```suggestion
	printf("Listed objects: %s\n", dsname);
```

File: cmd/uzfs/uzfs.c [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+
+int
+uzfs_object_read(int argc, char **argv)
+{
+	int err = 0;
+	char *dsname = argv[1];
+	uint64_t obj = atoi(argv[2]);
+	int offset = atoi(argv[3]);
+	int size = atoi(argv[4]);
+	char buf[1024] = "";
```
Comment (kylezh) [inline cmd/uzfs/uzfs.c:541]: just 1024?

File: cmd/uzfs/uzfs.c [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+
+int
+uzfs_object_write(int argc, char **argv)
+{
+	int err = 0;
+	char *dsname = argv[1];
+	uint64_t obj = atoi(argv[2]);
+	int offset = atoi(argv[3]);
+	int size = atoi(argv[4]);
+	char buf[1024] = "";
```
Comment (kylezh) [inline cmd/uzfs/uzfs.c:568]: ditto

File: lib/libuzfs/libuzfs.c [hunk 10]
```diff
@@ -0,0 +1,740 @@
...
+		if (ftruncate(fd, size) != 0)
+			fatal(1, "can't ftruncate %s", path);
+		(void) close(fd);
+	}
+
+	file = fnvlist_alloc();
+	fnvlist_add_string(file, ZPOOL_CONFIG_TYPE, VDEV_TYPE_FILE);
+	fnvlist_add_string(file, ZPOOL_CONFIG_PATH, path);
+	fnvlist_add_uint64(file, ZPOOL_CONFIG_ASHIFT, ashift);
+	umem_free(pathbuf, MAXPATHLEN);
```
Comment (kylezh) [inline lib/libuzfs/libuzfs.c:136]:
```suggestion

	umem_free(pathbuf, MAXPATHLEN);
```

File: lib/libuzfs/libuzfs.c [hunk 11]
```diff
@@ -0,0 +1,740 @@
...
+{
+	return (spa_destroy(zpool));
+}
+
+libuzfs_zpool_handle_t *
+libuzfs_zpool_open(const char *zpool)
+{
+	int err = 0;
+	spa_t *spa;
+	err = spa_open(zpool, &spa, FTAG);
```
Comment (kylezh) [inline lib/libuzfs/libuzfs.c:364]:
```suggestion

	err = spa_open(zpool, &spa, FTAG);
```

File: lib/libuzfs/libuzfs.c [hunk 12]
```diff
@@ -0,0 +1,740 @@
...
+
+	err = dmu_bonus_hold(os, obj, FTAG, &db);
+	if (err)
+		return (err);
+
+	dmu_object_info_from_db(db, doi);
+
+	dmu_buf_rele(db, FTAG);
+	return (0);
+
```
Comment (kylezh) [inline lib/libuzfs/libuzfs.c:566]:
```suggestion
```

File: lib/libuzfs/libuzfs.c [hunk 13]
```diff
@@ -0,0 +1,740 @@
...
+	return (err);
+}
+
+int
+libuzfs_object_claim(libuzfs_dataset_handle_t *dhp, uint64_t obj)
+{
+	int err = 0;
+	dmu_tx_t *tx;
+	dmu_buf_t *db;
+	objset_t *os = dhp->os;
```
Comment (kylezh) [inline lib/libuzfs/libuzfs.c:638]:
```suggestion
	int err = 0;
	dmu_tx_t *tx = NULL;
	dmu_buf_t *db = NULL;
	objset_t *os = dhp->os;
```

File: lib/libuzfs/libuzfs.c (hunk 3 above)
Comment (hpingfs) [reply lib/libuzfs/libuzfs.c:475]: find will call the parameter: libuzfs_objset_destroy_cb to do real work.

File: lib/libuzfs/libuzfs.c [hunk 14]
```diff
@@ -149,8 +150,8 @@ static nvlist_t *
 make_vdev_mirror(const char *path, char *aux, const char *pool, size_t size,
     uint64_t ashift, int r, int m)
 {
-	nvlist_t *mirror, **child;
-	int c;
+	int c = 0;
+	nvlist_t *mirror = NULL, **child = NULL;
```
Comment (kylezh) [inline lib/libuzfs/libuzfs.c:154]: Don't define multiple variables in one line.

File: lib/libuzfs/libuzfs.c [hunk 15]
```diff
@@ -492,7 +496,7 @@ libuzfs_dhp_fini(libuzfs_dataset_handle_t *dhp)
 static void
 libuzfs_dataset_dirobj_verify(libuzfs_dataset_handle_t *dhp)
 {
-	uint64_t usedobjs, dirobjs, scratch;
+	uint64_t usedobjs = 0, dirobjs = 0, scratch = 0;
```
Comment (kylezh) [inline lib/libuzfs/libuzfs.c:499]: ditto

**Comment by shuoranliu** [comment]
\LGTM

**Review by kylezh — APPROVED** [review]

**Pull request merged**